### PR TITLE
Added Batch Normalization Layer modules

### PR DESCRIPTION
**Pull request by Spnetic-5**
Addresses #155  

@milancurcic , I've included the structure of the batch normalization layer. Could you please review it and confirm whether I'm in the correct direction?

### CODE REVIEW

**Comment by milancurcic** [comment]
Thank you @Spnetic-5, great. Yes, this is certainly in the correct direction. I haven't yet reviewed the forward and backward algorithms, let me know when it's ready for review.

Also, despite #156, we should be able to test this layer implementation on its own, without integrating it with the network. In other words, while #156 will be necessary for full integration, we can work on this implementation as a standalone layer before #156.

**Comment by Spnetic-5** [comment]
@milancurcic  Please review the forward and backward pass implementations I've added based on my interpretation of the paper, also could you guide me on how we can test this layer?

**Comment by milancurcic** [comment]
Thanks! Let's review it briefly on the call today.

We can test this layer independently by passing some small, known input, and comparing the result with the corresponding known output. This should be straightforward since the batchnorm operation is relatively simple (just normalization of data). The backward is the inverse operation, so as I understand it, we can pass the same expected output to recover the same expected input.

**Comment by milancurcic** [comment]
@Spnetic-5 I just saw your message on Discourse, no problem; we'll proceed work on this PR as usual.

**Comment by milancurcic** [comment]
See for example a program that tests forward and backward passes of the maxpool2d layer using known inputs and expected outputs:

https://github.com/modern-fortran/neural-fortran/blob/main/test/test_maxpool2d_layer.f90

We'd use the same approach to test a batchnorm layer.

**Comment by Spnetic-5** [comment]
Hello, @milancurcic. Sorry for the lack of activity over the past few days; this was my final week of internship in Canada, and I'll be returning to India on Monday.

I added a test module for the batch norm layer, however it has some error; I believe I will need your assistance on this.

**Comment by milancurcic** [comment]
No worries at all, thanks for all the work. I'll review it tomorrow.

File: src/nf/nf_batchnorm_layer_submodule.f90 [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    integer, intent(in) :: num_features
+    type(batchnorm_layer) :: res
+
+    res % num_features = num_features
+    allocate(res % gamma(num_features), source=1.0)
+    allocate(res % beta(num_features))
+    allocate(res % running_mean(num_features), source=0.0)
+    allocate(res % running_var(num_features), source=1.0)
+    allocate(res % input(num_features, num_features))
+    allocate(res % output(num_features, num_features))
```
Comment (milancurcic) [inline src/nf/nf_batchnorm_layer_submodule.f90:18]: I don't think the shape for these should be `(num_features, num_features)`, but rather `(batch_size, num_features)`. The `batch_size` also won't be known until the first forward pass, so we should defer the allocation until then. In the forward pass, we could have a simple `allocated` check to see if they have not been allocated then, and allocate them to the shape of the `input`.

File: src/nf/nf_batchnorm_layer_submodule.f90 [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    class(batchnorm_layer), intent(in out) :: self
+    real, intent(in) :: input(:,:)
+
+    ! Store input for backward pass
+    self % input = input
+
+    associate( &
+      ! Normalize the input
+      normalized_input => (input - reshape(self % running_mean, shape(input, 1))) &
+        / sqrt(reshape(self % running_var, shape(input, 1)) + self % epsilon) &
```
Comment (milancurcic) [inline src/nf/nf_batchnorm_layer_submodule.f90:52]: `running_mean` and `running_var` are not yet updated anywhere, only initialized.

File: test/test_batchnorm_layer.f90 [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    write(stderr, '(a)') 'batchnorm layer should not be marked as initialized yet.. failed'
+  end if
+
+  ! Initialize sample input and gradient
+  allocate(sample_input(num_features, 1))
+  allocate(gradient(num_features, 1))
+  sample_input = 1.0
+  gradient = 2.0
+
+  !TODO run forward and backward passes directly on the batchnorm_layer instance
```
Comment (milancurcic) [inline test/test_batchnorm_layer.f90:37]: I'll add a simple test directly on the `batchnorm_layer` instance rather than the high-level `layer_type` instance so you get the idea how it will work.

File: src/nf/nf_batchnorm_layer_submodule.f90 (hunk 1 above)
Comment (milancurcic) [reply src/nf/nf_batchnorm_layer_submodule.f90:18]: Sorry, I meant `(num_features, batch_size)`.
